src/fsharp/FSharp.Core/prim-types.fs

```diff
@@ -744,7 +744,101 @@ namespace Microsoft.FSharp.Core
+                interface IEqualityComparer 
+                
+            /// The implementation of IEqualityComparer, using unlimited depth for hashing and PER semantics for NaN equality.
+            type UnlimitedHasherPER() =
+                interface IEqualityComparer
+                    
+
+            /// The unique object for unlimited depth for hashing and ER semantics for equality.
+            let fsEqualityComparerUnlimitedHashingER = UnlimitedHasherER()
+
+            /// The unique object for unlimited depth for hashing and PER semantics for equality.
+            let fsEqualityComparerUnlimitedHashingPER = UnlimitedHasherPER()
+                     
+            let inline HashCombine nr x y = (x <<< 1) + y + 631 * nr
+
+            let inline ArrayHashing<'element,'array when 'array :> System.Array> get lowerBound (f:'element->int) (x:'array) : int =
+                let rec loop acc i =
+                    if i < lowerBound then acc
+                    else loop (HashCombine i acc (f (get x i))) (i-1)
+                
+                let lastIdx =
+                    let upperBound = lowerBound+defaultHashNodes
+                    match lowerBound+x.Length-1 with
+                    | oversized when oversized > upperBound -> upperBound
+                    | good -> good
+
+                loop 0 lastIdx 
+
+            let GenericHashObjArray   (iec:System.Collections.IEqualityComparer) (x:obj[]) = ArrayHashing get 0 iec.GetHashCode x
+            let GenericHashByteArray  (x:byte[])  = ArrayHashing get 0 intOfByte x
+            let GenericHashInt32Array (x:int32[]) = ArrayHashing get 0 (fun x -> x) x
+            let GenericHashInt64Array (x:int64[]) = ArrayHashing get 0 int32 x
+
+            // special case - arrays do not by default have a decent structural hashing function
+            let GenericHashArbArray (iec : System.Collections.IEqualityComparer) (x: System.Array) : int =
+                match x.Rank with 
+                | 1 -> ArrayHashing (fun a i -> a.GetValue i) (x.GetLowerBound 0) iec.GetHashCode x
+                | _ -> HashCombine 10 (x.GetLength(0)) (x.GetLength(1)) 
+
+            // Core implementation of structural hashing, corresponds to pseudo-code in the 
+            // F# Language spec.  Searches for the IStructuralHash interface, otherwise uses GetHashCode().
+            // Arrays are structurally hashed through a separate technique.
+            //
+            // "iec" is either fsEqualityComparerUnlimitedHashingER, fsEqualityComparerUnlimitedHashingPER or a CountLimitedHasherPER.
+            let rec GenericHashParamObj (iec : System.Collections.IEqualityComparer) (x: obj) : int =
+                  match x with 
+                  | null -> 0 
+                  | (:? System.Array as a) -> 
+                      match a with 
+                      | :? (obj[]) as oa -> GenericHashObjArray iec oa 
+                      | :? (byte[]) as ba -> GenericHashByteArray ba 
```

Comment:
`sbyte[]` will match on this branch as well. Are you sure `GenericHashByteArray` will produce correct result in this case?

Reply:
yep, I know, still WIP...